diff --git a/TartanClasses.py b/TartanClasses.py
index 6d32c71..f64cca3 100644
--- a/TartanClasses.py
+++ b/TartanClasses.py
@@ -7789,7 +7789,10 @@ class ShowImage(object):
                 height = msiz
                 width = int(height / relate)
             self.currentImage = {}
-            image = pilimg.resize((width, height), Image.ANTIALIAS)
+            try:
+                image = pilimg.resize((width, height), Image.ANTIALIAS)
+            except:
+                image = pilimg.resize((width, height), Image.LANCZOS)
             self.currentImage["data"] = image
             self.frm = MyFrame(self.vbox, width=width, height=height)
             if self.vbox:
@@ -13950,7 +13953,7 @@ class RepPrt(object):
                 A combination of the two options.
     stots   :   A list of lists of columns, in least significant order, to be
                 breaks for sub-totalling as follows:
-                [] = No sub-totals
+                    [] = No sub-totals
                                         OR
                     column or "as" name
                     description or another column's contents
@@ -14492,18 +14495,32 @@ class RepPrt(object):
         self.grandTotal()
         try:
             if self.repprt[2] == "export":
+                ctots = []
+                #for tot in self.gtots:
+                #    cnt = 0
+                #    for col in self.cols:
+                #        if len(col) > 4 and col[4] == "N":
+                #            continue
+                #        if col[0].count(tot):
+                #            ctots.append(cnt)
+                #            break
+                #        cnt += 1
                 doWriteExport(rcdic=self.mf.rcdic, xtype=self.repprt[1],
                     name=self.expnam, heads=self.expheads, colsh=self.expcolsh,
-                    forms=self.expforms, datas=self.expdatas)
+                    forms=self.expforms, datas=self.expdatas, ctots=ctots)
             else:
+                if len(self.heads) > 1:
+                    rpthd = self.heads[1]
+                else:
+                    rpthd = self.heads[0]
                 if self.tails:
                     self.fpdf.setFont(style="B")
                     for tail in self.tails:
                         self.fpdf.drawText(txt=tail)
                 if self.sveprt and self.fpdf.saveFile(self.pdfnam):
                     doPrinter(mf=self.mf, conum=self.conum, pdfnam=self.pdfnam,
-                        header=self.heads[-1], fromad=self.fromad,
-                        repprt=self.repprt, repeml=self.repeml)
+                        header=rpthd, fromad=self.fromad, repprt=self.repprt,
+                        repeml=self.repeml)
         except:
             pass
 
@@ -17963,7 +17980,7 @@ Do you want to Export All Linked Companies?""" % (tx2, tx1), default="no")
                 continue
             self.doTable(tab)
             num += 1
-        self.only = True
+        self.only = False
         if 1 not in self.cono:
             self.only = True
         for tab in self.coys:
@@ -18959,10 +18976,10 @@ class ViewPDF(object):
                 self.links[link["from"]] = link["uri"]
         if not self.maxi and self.links:
             self.win.bind("<Button-1>", self.showLinks)
-            self.win.bind("<Motion>", self.changeCursor)
+            #self.win.bind("<Motion>", self.changeCursor)
         else:
             self.win.unbind("<Button-1>")
-            self.win.unbind("<Motion>")
+            #self.win.unbind("<Motion>")
         # Create image
         self.pgd.configure(state="normal")
         self.pgd.delete(0, "end")
diff --git a/bwl/bc2010.py b/bwl/bc2010.py
index 39f328b..0204562 100644
--- a/bwl/bc2010.py
+++ b/bwl/bc2010.py
@@ -1779,13 +1779,13 @@ Combination Number %10s"""
                     (self.count, self.tot)
             else:
                 if self.tsize == 3:
-                    bcl = "Pairs"
+                    btp = "Pairs"
                 else:
-                    bcl = "Broken"
+                    btp = "Broken"
                 txt = "Best Random Draw After Trying %s Different "\
                     "Combinations, Largest Team Difference is %s, "\
                     "Skips Clash %s, Players Clash %s, %s %s" % \
-                    (self.count, self.tot, self.scl, self.pcl, bcl, self.bcl)
+                    (self.count, self.tot, self.scl, self.pcl, btp, self.bcl)
             self.doShowDraw(txt, self.adraw1)
             self.dedit = "N"
             self.doSave()
@@ -2290,7 +2290,7 @@ Combination Number %10s"""
                 drw.append([f[0], nam, f[5]])
         self.adraw2.append(drw)
 
-    def doSave(self):
+    def doSave(self, stat=True):
         # Delete all existing records
         self.sql.delRec("bwldrm", where=[("bdm_cono", "=", self.opts["conum"]),
             ("bdm_date", "=", self.date), ("bdm_time", "=", self.time)])
@@ -2299,8 +2299,10 @@ Combination Number %10s"""
         # Insert bwldrm
         if self.stats:
             stat = self.stats
-        else:
+        elif self.drawn:
             stat = "%s %s %s %s" % (self.tot, self.scl, self.pcl, self.bcl)
+        else:
+            stat = ""
         self.sql.insRec("bwldrm", data=[self.opts["conum"], self.date,
             self.time, self.mixgd, self.mixrt, self.nbase, self.dtype,
             self.dhist, self.dedit, self.tsize, self.mrate.work,
diff --git a/bwl/bc2050.py b/bwl/bc2050.py
index 9452995..9491d39 100644
--- a/bwl/bc2050.py
+++ b/bwl/bc2050.py
@@ -487,7 +487,7 @@ class bc2050(object):
         whr = [
             ("bcg_cono", "=", self.opts["conum"]),
             ("bcg_ccod", "=", self.ccod)]
-        if self.cfmat in ("T", "X"):
+        if not self.reprint and self.cfmat in ("T", "X"):
             # Delete bwlgme records
             wer = copyList(whr)
             wer.append(("bcg_game", ">=", self.game))
diff --git a/bwl/bcc110.py b/bwl/bcc110.py
index 37ea762..4e78a7f 100644
--- a/bwl/bcc110.py
+++ b/bwl/bcc110.py
@@ -89,7 +89,7 @@ class bcc110(object):
         r1s = (("Yes","Y"),("No","N"))
         r2s = (("Position","P"),("Rating","R"),("Combined","C"))
         r3s = (("Ascending","A"),("Descending","D"))
-        r4s = (("S32L-S32L","A"),("S32L-L23S","B"))
+        r4s = (("S32L-L23S","B"),("S32L-S32L","A"))
         if self.memctl:
             self.fld = [
                 (("T",0,0,0),["IRB",r1s],0,"M/L Integration","",
diff --git a/doc/CTL.rst b/doc/CTL.rst
index 6c5607f..f099cee 100644
--- a/doc/CTL.rst
+++ b/doc/CTL.rst
@@ -9,7 +9,7 @@ This routine is used to create or amend a system control record by entering the
 + **Minimum Characters** - If you elected to enforce passwords enter the minimum number of characters that the passwords must consist of.
 + **Maximum Life (Days)** - If you elected to enforce passwords enter the maximum number of days that the passwords will be valid for.
 + **Backup History (Days)** - Enter the maximum number of days that backups will be retained. Enter 0 for infinity.
-+ **SMTP Server** - If you have email then enter your SMTP server's address e.g. smtp.mweb.co.za or smtp.saix.net or smtp.vodacom.co.za
++ **SMTP Server** - If you have email then enter your SMTP server's address e.g. smtp.mweb.co.za or smtp.saix.net or smtp.vodacom.co.za etc.
 + **SMTP Security** - Select the appropriate security method.
 + **SMTP Authentication** - Select the appropriate authentication method.
 + **SMTP Port Number** - Select the appropriate smtp port number.
@@ -30,7 +30,7 @@ While most of the fields are self explanatory the following are of note:
 + **E-Mail Address** - The company's default email address.
 + **V.A.T. Default** - Any alphanumeric character. If the code, except **N** which is used for no V.A.T., has not previously been created you will be prompted to create one as in `V.A.T. Records Maintenance`_.
 + **Systems** - You must tick all the systems that you would like to enable otherwise they will not appear on your menu.
-+ **Logo** - You can select an image file to be used as the letterhead for invoices, statements and remittance advices. this file should be a jpg file.
++ **Logo** - You can select an image file to be used as the letterhead for invoices, statements and remittance advices. This file should be a `jpeg` file.
 
 After creating a **new** company record you will be prompted for the initial financial period's starting and ending dates e.g. 20060301 and 20070228.
 
diff --git a/doc/SYS.rst b/doc/SYS.rst
index 9015d45..a7d8329 100644
--- a/doc/SYS.rst
+++ b/doc/SYS.rst
@@ -3,7 +3,7 @@ Tartan Systems - Reference Manual
 =================================
 .. _GPL: http://www.gnu.org/licenses/gpl.html
 
-:Version:   6.17
+:Version:   6.18
 :Author:    Paul Malherbe
 :Contact:   paul@tartan.co.za
 :Home:      http://www.tartan.co.za
diff --git a/gen/gl2010.py b/gen/gl2010.py
index 09c1bac..5dd3e98 100644
--- a/gen/gl2010.py
+++ b/gen/gl2010.py
@@ -73,7 +73,7 @@ Please Contact your Software Support and Quote this Message as this is a Serious
             (("C",0,0,0),"IUI",7,"Acc-Num","Account Number",
                 "","N",self.doAcc,sel_acc,None,None),
             (("C",0,0,1),"ONA",30,"Description"),
-            (("C",0,0,2),"ISD",13.2,"Balance","Balance Value",
+            (("C",0,0,2),"ISD",15.2,"Balance","Balance Value",
                 "","N",self.doBal,None,None,None))
         but = (
             ("Import File",None,self.doImport,0,("C",0,1),("C",0,2),
@@ -90,7 +90,7 @@ Please Contact your Software Support and Quote this Message as this is a Serious
         self.df.setWidget(self.df.mstFrame, state="hide")
         impcol = [
             ["Account Number", 0, "UI", 7],
-            ["Opening Balance", 1, "SD", 13.2]]
+            ["Opening Balance", 1, "SD", 15.2]]
         fi = FileImport(self.opts["mf"], impcol=impcol)
         err = None
         for num, line in enumerate(fi.impdat):
@@ -169,7 +169,7 @@ Please Contact your Software Support and Quote this Message as this is a Serious
             where=[("glo_cono", "=", self.opts["conum"]), ("glo_trdt", "=",
             self.start)], limit=1)
         if bals[0]:
-            diff = CCD(bals[0], "SD", 13.2)
+            diff = CCD(bals[0], "SD", 15.2)
             showError(self.opts["mf"].body, "Out of Balance",
                 "Opening Balances Do Not Balance by %s" % diff.disp)
             self.df.focusField(self.df.frt, self.df.pag, self.df.col)
diff --git a/gen/gl3030.py b/gen/gl3030.py
index fdf3ce1..d3a64a6 100644
--- a/gen/gl3030.py
+++ b/gen/gl3030.py
@@ -52,7 +52,7 @@ class gl3030(object):
             "Typ", "Batch", "Remarks", "Debit", "Credit", "Balance"]
         self.forms = [("UI", 3), ("UI", 7), ("NA", 30), ("D1", 10),
             ("Na", 9), ("NA", 3), ("Na", 7), ("NA", 30), ("SD", 13.2),
-            ("SD", 13.2), ("SD", 14.2)]
+            ("SD", 13.2), ("SD", 15.2)]
         self.s_per = int(self.opts["period"][1][0] / 100)
         self.e_per = int(self.opts["period"][2][0] / 100)
         self.recs = None
@@ -221,7 +221,7 @@ class gl3030(object):
     def printReport(self, recs):
         self.newpage = True
         p = ProgressBar(self.opts["mf"].body, mxs=len(recs), esc=True)
-        self.head = "%03u %-101s" % (self.opts["conum"], self.opts["conam"])
+        self.head = "%03u %-102s" % (self.opts["conum"], self.opts["conam"])
         self.fpdf = MyFpdf(name=self.__class__.__name__, head=self.head)
         for seq, dat in enumerate(recs):
             p.displayProgress(seq)
@@ -240,22 +240,22 @@ class gl3030(object):
                 else:
                     self.newAccount()
             if acctot:
-                w1 = CCD(acctot, "SD", 14.2)
+                w1 = CCD(acctot, "SD", 15.2)
                 self.fpdf.drawText("%-10s %-9s %-3s %-7s %-30s %-13s %-13s "\
-                    "%-14s" % (self.sdate.disp + "-01", "", "", "",
+                    "%-15s" % (self.sdate.disp + "-01", "", "", "",
                     "Opening Balance", "", "", w1.disp))
             for acc in trn:
                 trdt, refno, trtp, batch, amt, dbt, crt, detail, curdt, \
                     curmth = self.getTrnValues(acc)
                 if self.fpdf.newPage():
                     self.pageHeading()
-                    bf = CCD(acctot, "SD", 14.2)
+                    bf = CCD(acctot, "SD", 15.2)
                     if bf.work:
-                        self.fpdf.drawText("%32s %-30s %27s %14s" % ("",
+                        self.fpdf.drawText("%32s %-30s %27s %15s" % ("",
                             "Brought Forward", "", bf.disp))
                 acctot = float(ASD(acctot) + ASD(amt.work))
-                w1 = CCD(acctot, "SD", 14.2)
-                self.fpdf.drawText("%-10s %-9s %-3s %-7s %-30s %13s %13s %14s"\
+                w1 = CCD(acctot, "SD", 15.2)
+                self.fpdf.drawText("%-10s %-9s %-3s %-7s %-30s %13s %13s %15s"\
                     % (trdt.disp, refno.disp, gltrtp[(trtp.work - 1)][0],
                     batch.disp, detail.disp, dbt.disp, crt.disp, w1.disp))
             if self.pages == "Y" and (acctot or trn):
@@ -279,18 +279,18 @@ class gl3030(object):
             where=[("glo_cono", "=", self.opts["conum"]), ("glo_acno", "=",
             acno.work), ("glo_trdt", "=", self.opts["period"][1][0])], limit=1)
         if o:
-            b = CCD(o[0], "SD", 14.2)
+            b = CCD(o[0], "SD", 15.2)
         else:
-            b = CCD(0, "SD", 14.2)
+            b = CCD(0, "SD", 15.2)
         ob = b.work
         o = self.sql.getRec("gentrn", cols=["round(sum(glt_tramt), 2)"],
             where=[("glt_cono", "=", self.opts["conum"]), ("glt_acno", "=",
             acno.work), ("glt_curdt", ">=", self.s_per), ("glt_curdt",
             "<", self.sdate.work)], limit=1)
         if o and o[0]:
-            b = CCD(o[0], "SD", 14.2)
+            b = CCD(o[0], "SD", 15.2)
         else:
-            b = CCD(0, "SD", 14.2)
+            b = CCD(0, "SD", 15.2)
         acctot = float(ASD(ob) + ASD(b.work))
         return (acno, desc, acctot)
 
@@ -340,9 +340,9 @@ class gl3030(object):
             self.fpdf.drawText("%-7s %-7s %-30s" % ("Account", self.acno.disp,
                 self.desc.disp))
             self.fpdf.drawText()
-            self.fpdf.drawText("%-10s %-9s %-3s %-7s %-30s %-13s %-13s %-14s" \
+            self.fpdf.drawText("%-10s %-9s %-3s %-7s %-30s %-13s %-13s %-15s" \
                 % ("   Date", "Reference", "Typ", "Batch", "Remarks",
-                "       Debit", "      Credit", "     Balance"))
+                "       Debit", "      Credit", "       Balance"))
             self.fpdf.underLine(self.head)
             self.fpdf.setFont()
 
diff --git a/gen/gl3090.py b/gen/gl3090.py
index fba7343..2ca69f8 100644
--- a/gen/gl3090.py
+++ b/gen/gl3090.py
@@ -114,12 +114,13 @@ class gl3090(object):
             whr.append(("grt_date", "<=", self.to))
         if self.unall == "Y":
             whr.append(("grt_flag", "=", "N"))
-        odr = "grt_date, grt_refno"
+        tab = ["genrct"]
         tot = ["grt_amount"]
+        odr = "grt_date, grt_refno"
         RepPrt(self.opts["mf"], conum=self.opts["conum"],
-            conam=self.opts["conam"], name=self.__class__.__name__,
-            tables=["genrct"], heads=hds, cols=col, where=whr, order=odr,
-            gtots=tot, repprt=self.df.repprt, repeml=self.df.repeml)
+            conam=self.opts["conam"], name=self.__class__.__name__, tables=tab,
+            heads=hds, cols=col, where=whr, order=odr, gtots=tot,
+            repprt=self.df.repprt, repeml=self.df.repeml)
         self.closeProcess()
 
     def doExit(self):
diff --git a/gen/gl4010.py b/gen/gl4010.py
index 419d51e..d5d0d5b 100644
--- a/gen/gl4010.py
+++ b/gen/gl4010.py
@@ -106,11 +106,11 @@ class gl4010(object):
             (("T",0,0,0),"ONA",30,"Description"),
             (("T",0,0,0),"OUA",1,"Type"),
             (("T",0,0,0),"OUA",1,"Indicator"),
-            (("C",1,0,0),"OSD",14.2,"Actual","","","N",
+            (("C",1,0,0),"OSD",15.2,"Actual","","","N",
                 None,None,None,None,("Months",14)),
-            (("C",1,0,1),"OSD",14.2,"Budget"),
-            (("C",1,0,2),"OSD",14.2,"Variance"),
-            (("C",1,0,3),"OSD",14.2,"Last-Year"),
+            (("C",1,0,1),"OSD",15.2,"Budget"),
+            (("C",1,0,2),"OSD",15.2,"Variance"),
+            (("C",1,0,3),"OSD",15.2,"Last-Year"),
             (("T",2,0,0),"Id2",7,"Period (YYYYMM)","Financial Period (0=All)",
                 self.trnper,"Y",self.doPeriod,None,None,None))
         tag = (
@@ -276,7 +276,7 @@ class gl4010(object):
             ("glt_batch", "", 0, "Batch"),
             ("glt_tramt", "", 0, "       Debit"),
             ("glt_tramt", "", 0, "      Credit"),
-            ("balance", "SD", 14.2, "     Balance"),
+            ("balance", "SD", 15.2, "     Balance"),
             ("glt_desc", "", 0, "Remarks"),
             ("glt_seq", "", 0, "Sequence"))
         mthno, acctot = self.getObal()
@@ -432,7 +432,7 @@ class gl4010(object):
                     ["glt_batch", "Na", 7, "BatchNo"],
                     ["debit", "SD", 13.2, "       Debit "],
                     ["credit", "SD", 13.2, "      Credit "],
-                    ["balance", "SD", 14.2, "     Balance "],
+                    ["balance", "SD", 15.2, "     Balance "],
                     ["glt_desc", "NA", 30, "Remarks"]]
                 if acctot:
                     if acctot < 0:
diff --git a/gen/gl6030.py b/gen/gl6030.py
index 72869ee..738aa35 100644
--- a/gen/gl6030.py
+++ b/gen/gl6030.py
@@ -177,9 +177,9 @@ class gl6030(object):
                             ("glo_acno", "=", acc), ("glo_trdt", "=",
                             self.start)], limit=1)
                         if o:
-                            b = CCD(o[0], "SD", 14.2)
+                            b = CCD(o[0], "SD", 15.2)
                         else:
-                            b = CCD(0, "SD", 14.2)
+                            b = CCD(0, "SD", 15.2)
                         self.gar_bal = float(ASD(self.gar_bal) + ASD(b.work))
                         o = self.sql.getRec("gentrn",
                             cols=["round(sum(glt_tramt), 2)"],
@@ -188,9 +188,9 @@ class gl6030(object):
                             self.s_per), ("glt_curdt", "<=",
                             self.opts["period"])], limit=1)
                         if o and o[0]:
-                            b = CCD(float(o[0]), "SD", 14.2)
+                            b = CCD(float(o[0]), "SD", 15.2)
                         else:
-                            b = CCD(0, "SD", 14.2)
+                            b = CCD(0, "SD", 15.2)
                         self.gar_bal = float(ASD(self.gar_bal) + ASD(b.work))
                         done.append(acc)
             o = self.sql.getRec("asstrn",
@@ -210,9 +210,9 @@ class gl6030(object):
                 where=[("glo_cono", "=", self.opts["conum"]), ("glo_acno", "=",
                 self.bkm_ctl), ("glo_trdt", "=", self.start)], limit=1)
             if o:
-                b = CCD(o[0], "SD", 14.2)
+                b = CCD(o[0], "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.gbk_bal = float(ASD(self.gbk_bal) + ASD(b.work))
             o = self.sql.getRec("gentrn",
                 cols=["round(sum(glt_tramt), 2)"], where=[("glt_cono", "=",
@@ -220,9 +220,9 @@ class gl6030(object):
                 ("glt_curdt", ">=", self.s_per), ("glt_curdt", "<=",
                 self.opts["period"])], limit=1)
             if o and o[0]:
-                b = CCD(float(o[0]), "SD", 14.2)
+                b = CCD(float(o[0]), "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.gbk_bal = float(ASD(self.gbk_bal) + ASD(b.work))
             bal = self.sql.getRec("bkmtrn",
                 cols=["round(sum(bkt_tramt), 2)"], where=[("bkt_cono", "=",
@@ -241,9 +241,9 @@ class gl6030(object):
                 where=[("glo_cono", "=", self.opts["conum"]), ("glo_acno", "=",
                 self.crs_ctl), ("glo_trdt", "=", self.start)], limit=1)
             if o:
-                b = CCD(o[0], "SD", 14.2)
+                b = CCD(o[0], "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.gcr_bal = float(ASD(self.gcr_bal) + ASD(b.work))
             o = self.sql.getRec("gentrn",
                 cols=["round(sum(glt_tramt), 2)"], where=[("glt_cono", "=",
@@ -251,9 +251,9 @@ class gl6030(object):
                 ("glt_curdt", ">=", self.s_per), ("glt_curdt", "<=",
                 self.opts["period"])], limit=1)
             if o and o[0]:
-                b = CCD(float(o[0]), "SD", 14.2)
+                b = CCD(float(o[0]), "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.gcr_bal = float(ASD(self.gcr_bal) + ASD(b.work))
             bal = self.sql.getRec("crstrn",
                 cols=["round(sum(crt_tramt), 2)"], where=[("crt_cono", "=",
@@ -272,9 +272,9 @@ class gl6030(object):
                 where=[("glo_cono", "=", self.opts["conum"]), ("glo_acno", "=",
                 self.drs_ctl), ("glo_trdt", "=", self.start)], limit=1)
             if o:
-                b = CCD(o[0], "SD", 14.2)
+                b = CCD(o[0], "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.gdr_bal = float(ASD(self.gdr_bal) + ASD(b.work))
             o = self.sql.getRec("gentrn",
                 cols=["round(sum(glt_tramt), 2)"], where=[("glt_cono", "=",
@@ -282,9 +282,9 @@ class gl6030(object):
                 ("glt_curdt", ">=", self.s_per), ("glt_curdt", "<=",
                 self.opts["period"])], limit=1)
             if o and o[0]:
-                b = CCD(float(o[0]), "SD", 14.2)
+                b = CCD(float(o[0]), "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.gdr_bal = float(ASD(self.gdr_bal) + ASD(b.work))
             bal = self.sql.getRec("drstrn",
                 cols=["round(sum(drt_tramt), 2)"], where=[("drt_cono", "=",
@@ -303,9 +303,9 @@ class gl6030(object):
                 where=[("glo_cono", "=", self.opts["conum"]), ("glo_acno", "=",
                 self.lon_ctl), ("glo_trdt", "=", self.start)], limit=1)
             if o:
-                b = CCD(o[0], "SD", 14.2)
+                b = CCD(o[0], "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.gln_bal = float(ASD(self.gln_bal) + ASD(b.work))
             o = self.sql.getRec("gentrn",
                 cols=["round(sum(glt_tramt), 2)"], where=[("glt_cono", "=",
@@ -313,9 +313,9 @@ class gl6030(object):
                 ("glt_curdt", ">=", self.s_per), ("glt_curdt", "<=",
                 self.opts["period"])], limit=1)
             if o and o[0]:
-                b = CCD(float(o[0]), "SD", 14.2)
+                b = CCD(float(o[0]), "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.gln_bal = float(ASD(self.gln_bal) + ASD(b.work))
             bal = self.sql.getRec("lontrn",
                 cols=["round(sum(lnt_tramt), 2)"], where=[("lnt_cono", "=",
@@ -334,9 +334,9 @@ class gl6030(object):
                 where=[("glo_cono", "=", self.opts["conum"]), ("glo_acno", "=",
                 self.mem_ctl), ("glo_trdt", "=", self.start)], limit=1)
             if o:
-                b = CCD(o[0], "SD", 14.2)
+                b = CCD(o[0], "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.gml_bal = float(ASD(self.gml_bal) + ASD(b.work))
             o = self.sql.getRec("gentrn",
                 cols=["round(sum(glt_tramt), 2)"], where=[("glt_cono", "=",
@@ -344,9 +344,9 @@ class gl6030(object):
                 ("glt_curdt", ">=", self.s_per), ("glt_curdt", "<=",
                 self.opts["period"])], limit=1)
             if o and o[0]:
-                b = CCD(float(o[0]), "SD", 14.2)
+                b = CCD(float(o[0]), "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.gml_bal = float(ASD(self.gml_bal) + ASD(b.work))
             bal = self.sql.getRec("memtrn",
                 cols=["round(sum(mlt_tramt), 2)"], where=[("mlt_cono", "=",
@@ -366,9 +366,9 @@ class gl6030(object):
                 ("glo_acno", "=", self.own_ctl), ("glo_trdt", "=",
                 self.start)], limit=1)
             if o:
-                b = CCD(o[0], "SD", 14.2)
+                b = CCD(o[0], "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.grc_own = float(ASD(self.grc_own) + ASD(b.work))
             o = self.sql.getRec("gentrn",
                 cols=["round(sum(glt_tramt), 2)"],
@@ -377,9 +377,9 @@ class gl6030(object):
                 self.s_per), ("glt_curdt", "<=", self.opts["period"])],
                 limit=1)
             if o and o[0]:
-                b = CCD(float(o[0]), "SD", 14.2)
+                b = CCD(float(o[0]), "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.grc_own = float(ASD(self.grc_own) + ASD(b.work))
             o = self.sql.getRec("rcaowt",
                 cols=["round(sum(rot_tramt), 2)"], where=[("rot_cono", "=",
@@ -395,9 +395,9 @@ class gl6030(object):
                 ("glo_acno", "=", self.tnt_ctl), ("glo_trdt", "=",
                 self.start)], limit=1)
             if o:
-                b = CCD(o[0], "SD", 14.2)
+                b = CCD(o[0], "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.grc_tnt = float(ASD(self.grc_tnt) + ASD(b.work))
             o = self.sql.getRec("gentrn",
                 cols=["round(sum(glt_tramt), 2)"],
@@ -406,9 +406,9 @@ class gl6030(object):
                 self.s_per), ("glt_curdt", "<=", self.opts["period"])],
                 limit=1)
             if o and o[0]:
-                b = CCD(float(o[0]), "SD", 14.2)
+                b = CCD(float(o[0]), "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.grc_tnt = float(ASD(self.grc_tnt) + ASD(b.work))
             o = self.sql.getRec("rcatnt",
                 cols=["round(sum(rtu_tramt), 2)"], where=[("rtu_cono", "=",
@@ -424,9 +424,9 @@ class gl6030(object):
                 ("glo_acno", "=", self.dep_ctl), ("glo_trdt", "=",
                 self.start)], limit=1)
             if o:
-                b = CCD(o[0], "SD", 14.2)
+                b = CCD(o[0], "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.grc_dep = float(ASD(self.grc_dep) + ASD(b.work))
             o = self.sql.getRec("gentrn",
                 cols=["round(sum(glt_tramt), 2)"],
@@ -435,9 +435,9 @@ class gl6030(object):
                 self.s_per), ("glt_curdt", "<=", self.opts["period"])],
                 limit=1)
             if o and o[0]:
-                b = CCD(float(o[0]), "SD", 14.2)
+                b = CCD(float(o[0]), "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.grc_dep = float(ASD(self.grc_dep) + ASD(b.work))
             o = self.sql.getRec("rcatnt",
                 cols=["round(sum(rtu_tramt), 2)"], where=[("rtu_cono", "=",
@@ -468,9 +468,9 @@ class gl6030(object):
                         ("glo_acno", "=", acc[0]), ("glo_trdt", "=",
                         self.start)], limit=1)
                     if o:
-                        b = CCD(o[0], "SD", 14.2)
+                        b = CCD(o[0], "SD", 15.2)
                     else:
-                        b = CCD(0, "SD", 14.2)
+                        b = CCD(0, "SD", 15.2)
                     self.grt_bal = float(ASD(self.grt_bal) + ASD(b.work))
                     o = self.sql.getRec("gentrn",
                         cols=["round(sum(glt_tramt), 2)"],
@@ -479,9 +479,9 @@ class gl6030(object):
                         self.s_per), ("glt_curdt", "<=", self.opts["period"])],
                         limit=1)
                     if o and o[0]:
-                        b = CCD(float(o[0]), "SD", 14.2)
+                        b = CCD(float(o[0]), "SD", 15.2)
                     else:
-                        b = CCD(0, "SD", 14.2)
+                        b = CCD(0, "SD", 15.2)
                     self.grt_bal = float(ASD(self.grt_bal) + ASD(b.work))
                     done.append(acc[0])
             o = self.sql.getRec("rtltrn",
@@ -501,9 +501,9 @@ class gl6030(object):
                 where=[("glo_cono", "=", self.opts["conum"]), ("glo_acno", "=",
                 self.str_ctl), ("glo_trdt", "=", self.start)], limit=1)
             if o:
-                b = CCD(o[0], "SD", 14.2)
+                b = CCD(o[0], "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.gst_bal = float(ASD(self.gst_bal) + ASD(b.work))
             o = self.sql.getRec("gentrn",
                 cols=["round(sum(glt_tramt), 2)"], where=[("glt_cono", "=",
@@ -511,9 +511,9 @@ class gl6030(object):
                 ("glt_curdt", ">=", self.s_per), ("glt_curdt", "<=",
                 self.opts["period"])], limit=1)
             if o and o[0]:
-                b = CCD(float(o[0]), "SD", 14.2)
+                b = CCD(float(o[0]), "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.gst_bal = float(ASD(self.gst_bal) + ASD(b.work))
             o = self.sql.getRec("strtrn",
                 cols=["round(sum(stt_cost), 2)"], where=[("stt_cono", "=",
@@ -532,9 +532,9 @@ class gl6030(object):
                 where=[("glo_cono", "=", self.opts["conum"]), ("glo_acno", "=",
                 self.sln_ctl), ("glo_trdt", "=", self.start)], limit=1)
             if o:
-                b = CCD(o[0], "SD", 14.2)
+                b = CCD(o[0], "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.gsl_bal = float(ASD(self.gsl_bal) + ASD(b.work))
             o = self.sql.getRec("gentrn",
                 cols=["round(sum(glt_tramt), 2)"], where=[("glt_cono", "=",
@@ -542,9 +542,9 @@ class gl6030(object):
                 ("glt_curdt", ">=", self.s_per), ("glt_curdt", "<=",
                 self.opts["period"])], limit=1)
             if o and o[0]:
-                b = CCD(float(o[0]), "SD", 14.2)
+                b = CCD(float(o[0]), "SD", 15.2)
             else:
-                b = CCD(0, "SD", 14.2)
+                b = CCD(0, "SD", 15.2)
             self.gsl_bal = float(ASD(self.gsl_bal) + ASD(b.work))
             o = self.sql.getRec("wagltf",
                 cols=["round(sum(wlt_amt), 2)"], where=[("wlt_cono", "=",
diff --git a/ms0000.py b/ms0000.py
index 15f18cd..4962bdd 100755
--- a/ms0000.py
+++ b/ms0000.py
@@ -65,7 +65,7 @@ if "TARVER" in os.environ:
     temp = tuple(os.environ["TARVER"].split("."))
     VERSION = (int(temp[0]), int(temp[1].rstrip()))
 else:
-    VERSION = (6, 17)
+    VERSION = (6, 18)
     os.environ["TARVER"] = "%s.%s" % VERSION
 
 class ms0000(object):
@@ -1655,40 +1655,39 @@ System --> Change Password""")
             dft = None
             if "wrka" in self.mf.rcdic and self.mf.rcdic["wrka"] == "Y":
                 ask = self.mf.rcdic["wrkf"]
+                if ask == "T" and not TRASH:
+                    ask = "D"
             else:
-                if "wrkf" in self.mf.rcdic:
-                    if self.mf.rcdic["wrkf"] == "T":
-                        dft = "Trash"
-                    elif self.mf.rcdic["wrkf"] == "D":
-                        dft = "Delete"
-                    else:
-                        dft = "Keep"
                 if TRASH:
                     but = [("Trash", "T", "Send files to Recycle Bin")]
-                    if not dft:
-                        dft = "Trash"
                 else:
                     but = []
-                if  not dft:
-                        dft = "Delete"
                 but.extend([
                     ("Delete", "D", "Permanently Delete the files"),
                     ("Keep", "K", "Keep the files in the Work Directory")])
+                if "wrkf" in self.mf.rcdic:
+                    if self.mf.rcdic["wrkf"] == "T":
+                        if TRASH:
+                            dft = "Trash"
+                        else:
+                            dft = "Delete"
+                    elif self.mf.rcdic["wrkf"] == "D":
+                        dft = "Delete"
+                    else:
+                        dft = "Keep"
+                if not dft:
+                    dft = "Delete"
                 ask = askChoice(self.mf.body, "Temporary Files",
                     "What do you wish to do with the Temporary Report Files "\
                     "in the wrk Directory?", butt=but, default=dft)
-            if ask == "T":
-                for fle in fles:
-                    try:
+            for fle in fles:
+                try:
+                    if ask == "T":
                         send2trash(fle)
-                    except:
-                        pass
-            elif ask == "D":
-                for fle in fles:
-                    try:
+                    elif ask == "D":
                         os.remove(fle)
-                    except:
-                        pass
+                except:
+                    pass
 
     def doExit(self, dbm=True):
         if dbm and self.dbm.dbopen:
diff --git a/mst/ms1010.py b/mst/ms1010.py
index a05355c..cd5794c 100644
--- a/mst/ms1010.py
+++ b/mst/ms1010.py
@@ -266,7 +266,7 @@ class ms1010(object):
                     pass
             self.img = ShowImage(self.df.nb.Page4, logo,
                 wrkdir=self.opts["mf"].rcdic["wrkdir"], msiz=640)
-        except:
+        except Exception as err:
             return "error"
 
     def doReadVat(self, w):
diff --git a/sls/si3030.py b/sls/si3030.py
index c88f31c..a4bc023 100644
--- a/sls/si3030.py
+++ b/sls/si3030.py
@@ -412,12 +412,11 @@ class si3030(object):
                 sveprt = True
             else:
                 sveprt = False
-            rp = RepPrt(self.opts["mf"], name=self.__class__.__name__,
+            RepPrt(self.opts["mf"], name=self.__class__.__name__,
                 tables=data, ttype="D", heads=heds, cols=cols,
                 stots=[["a", "Group Total", "Y"]], gtots=gtots,
                 repprt=self.df.repprt, repeml=self.df.repeml,
                 sveprt=sveprt)
-            self.fpdf = rp.fpdf
         self.opts["mf"].closeLoop()
 
     def doExit(self):
diff --git a/sls/si3090.py b/sls/si3090.py
index 8918179..444ca3e 100644
--- a/sls/si3090.py
+++ b/sls/si3090.py
@@ -96,7 +96,8 @@ class si3090(object):
         for c in col[:5]:
             cols.append((c, dics[c][2], dics[c][3], dics[c][5]))
         RepPrt(self.opts["mf"], name=self.__class__.__name__, tables=data,
-            ttype="D", heads=hed, cols=cols)
+            ttype="D", heads=hed, cols=cols, repprt=self.df.repprt,
+            repeml=self.df.repeml)
         self.opts["mf"].closeLoop()
 
     def doExit(self):
diff --git a/tab/tb1020.py b/tab/tb1020.py
index 01546fb..253c69e 100644
--- a/tab/tb1020.py
+++ b/tab/tb1020.py
@@ -174,6 +174,14 @@ class tb1020(object):
         for num, self.table in enumerate(self.tabs):
             if self.opts["bar"] and self.opts["mf"] and self.opts["mf"].window:
                 self.p1.displayProgress(num)
+            if self.table == "gentrn":
+                recon = self.sql.getRec("ffield", cols=["ff_type"],
+                    where=[("ff_tabl", "=", "gentrn"), ("ff_name", "=",
+                    "glt_recon")], limit=1)
+                if recon[0] == "UI":
+                    self.sql.updRec("ffield", cols=["ff_type"], data=["d2"],
+                    where=[("ff_tabl", "=", "gentrn"), ("ff_name", "=",
+                    "glt_recon")])
             recs = False
             chk = self.doReadTables("idx")
             if not chk:
diff --git a/tartanFunctions.py b/tartanFunctions.py
index 61301f6..a079062 100644
--- a/tartanFunctions.py
+++ b/tartanFunctions.py
@@ -2376,7 +2376,7 @@ def getFileName(path, ptyp="FF", wrkdir=None, check=False):
             os.chmod(nam, stat.S_IRWXU)
         con.close()
         return nam
-    except:
+    except Exception as err:
         if con:
             con.close()
         if fle:
@@ -2786,6 +2786,7 @@ def doWriteExport(**args):
                         "ULINES" - A sigle underline - skip
                         "TOTAL"  - A total line of columns of data
                     List of values
+        ctots  - A list of TOTAL columns to sum using the SUBTOTAL function
         rcdic  - The tartanrc dictionary
         view   - Whether to view the report
         wait   - Whether to wait for the viewer to exit.
@@ -2938,10 +2939,10 @@ def doWriteExport(**args):
         blank = False
         for num, valx in enumerate(args["heads"]):
             if type(valx) in (list, tuple):
-                cols = len(args["colsh"][-1])
                 hgt, siz = valx[1:]
                 hxf = {"font": ["Ariel", siz, True]}
                 hxf["align"] = ["center", "center"]
+                cols = len(args["colsh"][-1])
             else:
                 hxf = {"font": "Ariel"}
                 hxf["bold"] = True
@@ -3036,6 +3037,13 @@ def doWriteExport(**args):
             sheet, rowx = createSheet(fmt, page)
             page += 1
         # Generate the body
+        if "ctots" in args and args["ctots"]:
+            gtot = {}
+            stot = {}
+            for tot in args["ctots"]:
+                if type(tot) in (list, tuple):
+                    stot[tot[0]]
+                stot[tot] = []
         for num, row in enumerate(args["datas"]):
             if row[0] == "PAGE":
                 args["heads"] = row[1][0]
@@ -3068,10 +3076,7 @@ def doWriteExport(**args):
                     unl = "d"
             # Write columns
             for colx, valx in enumerate(row[1]):
-                if type(valx) in (list, tuple):
-                    valx, bord = valx
-                else:
-                    bord = None
+                # Dates
                 if row[0] == "BODY":
                     if valx and args["forms"][colx][0].lower() == "d1":
                         # D1 Date conversion
@@ -3079,13 +3084,37 @@ def doWriteExport(**args):
                     elif valx and args["forms"][colx][0].lower() == "d2":
                         # D2 Date conversion
                         valx = datetime.datetime.strptime(str(valx), "%Y%m")
+                # Sum Totals
+                if "ctots" in args and args["ctots"]:
+                    if colx in stot:
+                        cola = getLetter(colx + 1)
+                        if row[0] == "BODY":
+                            if not stot[colx]:
+                                stot[colx] = ["%s%s" % (cola, rowx + 1),
+                                    "%s%s" % (cola, rowx + 1)]
+                                if not colx in gtot:
+                                    gtot[colx] = ["%s%s" % (cola, rowx + 1), ""]
+                            else:
+                                stot[colx][1] = "%s%s" % (cola, rowx + 1)
+                        elif row[0] == "TOTAL":
+                            if stot[colx]:
+                                valx = "=SUBTOTAL(9,%s:%s)" % tuple(stot[colx])
+                                gtot[colx][1] = "%s%s" % (cola, rowx + 1)
+                                stot[colx] = []
+                            elif gtot[colx]:
+                                valx = "=SUBTOTAL(9,%s:%s)" % tuple(gtot[colx])
                 # Format Font, Bold and/or Border
+                if type(valx) in (list, tuple):
+                    valx, bord = valx
+                else:
+                    bord = None
                 hxf = dict(fmt)
                 if row[0] in ("HEAD", "TOTAL"):
                     hxf["bold"] = True
                 # Underline & Border
                 ccd = list(args["forms"][colx][:])
-                if ccd[0] in numer and type(valx) == str:
+                if ccd[0] in numer and type(valx) == str and \
+                        not valx.count("=SUBTOTAL"):
                     ccd[0] = "NA"
                 if unl == "s" and ccd[0] in numer:
                     hxf["border"] = ["B", "thin"]
diff --git a/tartanWork.py b/tartanWork.py
index e6835d9..3d45087 100644
--- a/tartanWork.py
+++ b/tartanWork.py
@@ -667,6 +667,16 @@ payeRates = {
             [0, 0, 18]],
         "allow": [16425, 9000, 2997, 0]},
     "2024": {
+        "rates": [
+            [1817000, 644489, 45],
+            [857900, 251258, 41],
+            [673000, 179147, 39],
+            [512800, 121475, 36],
+            [370500, 77362, 31],
+            [237100, 42678, 26],
+            [0, 0, 18]],
+        "allow": [17235, 9444, 3145, 0]},
+    "2025": {
         "rates": [
             [1817000, 644489, 45],
             [857900, 251258, 41],
@@ -697,7 +707,8 @@ siteRates = {
     "2021": 0,
     "2022": 0,
     "2023": 0,
-    "2024": 0}
+    "2024": 0,
+    "2025": 0}
 uifRates = {
     "2004": [1.00, 1.00],
     "2005": [1.00, 1.00],
@@ -719,7 +730,8 @@ uifRates = {
     "2021": [1.00, 1.00],
     "2022": [1.00, 1.00],
     "2023": [1.00, 1.00],
-    "2024": [1.00, 1.00]}
+    "2024": [1.00, 1.00],
+    "2025": [1.00, 1.00]}
 sdlRates = {
     "2004": [0.00, 1.00],
     "2005": [0.00, 1.00],
@@ -741,7 +753,8 @@ sdlRates = {
     "2021": [0.00, 1.00],
     "2022": [0.00, 1.00],
     "2023": [0.00, 1.00],
-    "2024": [0.00, 1.00]}
+    "2024": [0.00, 1.00],
+    "2025": [0.00, 1.00]}
 # Labels
 labels = {
     "L7159": {
@@ -2258,7 +2271,7 @@ tabdic = {
             [0, "glo_cono", "UI", 3.0, "Company Number", "Coy"],
             [1, "glo_acno", "UI", 7.0, "Account Number", "Acc-Num"],
             [2, "glo_trdt", "D1", 10.0, "Year Start Date", "Start-Date"],
-            [3, "glo_cyr", "SD", 13.2, "Opening Balance - Amount", "Op-Amount"],
+            [3, "glo_cyr", "SD", 15.2, "Opening Balance - Amount", "Op-Amount"],
             [4, "glo_xflag", "UA", 1.0, "Export Flag", "X"]],
         "idx": [
             ["General Ledger Opening Bals", 1, "U", "glo_cono", "glo_acno",
@@ -2484,7 +2497,7 @@ tabdic = {
             [9, "glt_desc", "NA", 30.0, "Description", "Description"],
             [10, "glt_taxind", "UA", 1.0, "Tax Indicator", "V"],
             [11, "glt_batind", "UA", 1.0, "Batch Indicator", "B"],
-            [12, "glt_recon", "UI", 7.0, "Bank Date", "Bank-Dt"],
+            [12, "glt_recon", "d2", 7.0, "Bank Date", "Bank-Dt"],
             [13, "glt_capnm", "TX", 20.0, "Capture Name", "Capture-Name"],
             [14, "glt_capdt", "d1", 10.0, "Capture Date", "Capture-Dt"],
             [15, "glt_xflag", "UA", 1.0, "Export Flag", "X"],
